File: python-gdb-activities-20/activity18/gdb/logging/transaction_logger.py
# gdb/logging/transaction_logger.py
from typing import List
from gdb.domain.transaction import Transaction
from gdb.logging.log_destination import LogDestination

class TransactionLogger:
    """Bridge Abstraction: delegates logging operations to LogDestination implementor."""

    # ============================================================
    # 📝 STEP 4: Implement Bridge Constructor & Methods
    #
    # INSTRUCTIONS:
    #   1. __init__: store destination (a LogDestination) in self._destination.
    #   2. log(transaction): if transaction is not None, call self._destination.write(transaction).
    #   3. get_transactions(): return self._destination.read_all().
    #   4. get_destination(): return the current destination.
    #   5. set_destination(destination): replace self._destination -- this is how the backend is switched at runtime.
    # ============================================================
    # TODO: store the destination
    def __init__(self, destination: LogDestination) -> None:
        raise NotImplementedError("TODO: Step 4 - implement TransactionLogger constructor")

    # TODO: implement log, get_transactions, get_destination and set_destination
    def log(self, transaction: Transaction) -> None:
        pass

    def get_transactions(self) -> List[Transaction]:
        return []

    def get_destination(self) -> LogDestination:
        raise NotImplementedError("TODO: Step 4 - implement get_destination")

    def set_destination(self, destination: LogDestination) -> None:
        raise NotImplementedError("TODO: Step 4 - implement set_destination")

File: java-gdb-activities-20/activity18/src/com/gdb/logging/TransactionLogger.java
package com.gdb.logging;

import com.gdb.command.TransactionCommand;
import java.util.ArrayList;
import java.util.List;

/**
 * Bridge abstraction class decoupling high-level logger API from pluggable storage backends.
 */
public class TransactionLogger {
    // ============================================================
    // 📝 STEP 17: Declare Field
    // ============================================================
    // TODO: declare protected LogDestination destination;

    // ============================================================
    // 📝 STEP 18: Constructor
    // ============================================================
    // TODO: implement constructor accepting LogDestination
    public TransactionLogger(LogDestination destination) {
        // TODO: Step 18 - implement constructor
    }

    // ============================================================
    // 📝 STEP 19: setDestination
    // ============================================================
    // TODO: implement setter for hot-swapping destination
    public void setDestination(LogDestination destination) {
        // TODO: Step 19 - implement setter
    }

    // ============================================================
    // 📝 STEP 20: log(TransactionCommand cmd)
    // ============================================================
    // TODO: delegate to destination.write(cmd)
    public void log(TransactionCommand cmd) {
        // TODO: Step 20 - delegate to destination.write(cmd)
    }

    // ============================================================
    // 📝 STEP 21: readAll()
    // ============================================================
    // TODO: delegate to destination.readAll()
    public List<TransactionCommand> readAll() {
        // TODO: Step 21 - delegate to destination.readAll()
        return new ArrayList<>();
    }

    // ============================================================
    // 📝 STEP 22: clear()
    // ============================================================
    // TODO: delegate to destination.clear()
    public void clear() {
        // TODO: Step 22 - delegate to destination.clear()
    }

    // ============================================================
    // 📝 STEP 23: getDestinationName()
    // ============================================================
    // TODO: delegate to destination.getDestinationName()
    public String getDestinationName() {
        // TODO: Step 23 - delegate to destination.getDestinationName()
        return "";
    }
}
